contributing/architecture-eval-lifecycle.md

```diff
@@ -0,0 +1,326 @@
+
+    leaderRpc -->> httpAPI: Job.Register response
+    deactivate leaderRpc
+
+    httpAPI -->> cli: Create Job response
+    deactivate cli
+
+```
+
+## Scheduling
+
+A goroutine on the Nomad leader called the Eval Broker runs a continuous
+blocking query for Evaluations on the state store. Those Evaluations are
+enqueued in the broker.
+
+Schedulers are goroutines running on all server nodes. Typically this will be
+one per core on followers and 1/4 that number on the leader. The schedulers poll
+for Evaluations from the Eval Broker with the `Eval.Dequeue` RPC.
+
+Once a scheduler has an evaluation, it takes a snapshot of that server node's
+state store so that it has a consistent current view of the cluster state. The
+scheduler executes 3 main steps:
+
+* Reconcile: compare the cluster state and job specification to determine what
+  changes need to be made -- starting or stopping Allocations. The scheduler
+  creates new Allocations in this step. But note that _these Allocations are not
+  yet persisted to the state store_.
+* Feasibility Check: For each Allocation it needs, the scheduler iterates over
+  Nodes until it finds 2 Nodes that match the Allocation's resource requirements
+  and constraints.
+  * Note: for system jobs or jobs with with `spread` blocks, the scheduler has
+    to check all Nodes.
+* Scoring: For each feasible node, the scheduler ranks them and picks the one
+  with the highest score.
+
+If the job is a `service` job, the scheduler will also create (or update) a
+Deployment. When the scheduler determines the number of Allocations to create,
+it examine the job's [`update`][] block. Only the number of Allocations needed
+to complete the next phase of the update will be created in a given
+allocation. The Deployment is used by the deployment watcher (see below) to
+monitor the health of allocations and create new Evaluations to continue the
+update.
+
+If the scheduler cannot place all Allocations, it will create a new Evaluation
+in the "blocked" state and submit it to the leader. The Eval Broker will
+re-enque that Evaluation once cluster state has changed. (This process is the
+first green box in the sequence diagram below.)
+
+Once the scheduler has completed processing the Evaluation, if there are
+Allocations (and possibly a Deployment) to update, it will submit this work as a
+Plan to the leader. The leader needs to validate this plan and serialize it:
+
+* The scheduler took a snapshot of cluster state at the start of its work, so
+  that state may have changed in the meantime.
+* Schedulers run concurrently across the cluster, so they may generate
+  conflicting plans (particularly on heavily-packed clusters).
+
+The leader processes the plan in the plan applier. If the plan is valid, the
+plan applier will write the Allocations (and Deployment) update to the state
+store. If not, it will reject the plan. (The plan submit process is the second
+green box in the sequence diagram below.)
```

Comment:
```suggestion
store. If not, it will reject the plan and the scheduler will try to create a new plan with a refreshed state. If the scheduler fails to submit a valid plan too many times it submits a `blocked` Evaluation that is triggered by `max-plan-attempts` type. (The plan submit process is the second
green box in the sequence diagram below.)
```

I don't know if this helps to explain the `plan for node rejected` class of problems. Feel free to ignore if it's too much detail.

Another detail here is that the plan applier can have a partial apply, and so the scheduler only needs to retry what's left. But again, it may be too much to add here.